src/qibolab/instruments/qm.py

```diff
@@ -337,45 +339,78 @@ def register_integration_weights(self, qubit, readout_len):
         )
 
 
+@dataclass
+class AcquisitionVariables:
+    """QUA variables used for saving of acquisition results.
+    This class can be instantiated only within a QUA program scope.
+    Each readout pulse is associated with its own set of acquisition variables.
+    """
+
+    I: _Variable = field(default_factory=lambda: declare(fixed))
+    Q: _Variable = field(default_factory=lambda: declare(fixed))
+    """Variables to save the (I, Q) values acquired from a single shot."""
+    I_stream: _ResultSource = field(default_factory=lambda: declare_stream())
+    Q_stream: _ResultSource = field(default_factory=lambda: declare_stream())
+    """Streams to collect the results of all shots."""
+
+    raw_adc: bool = False
+    """Flag to select whether we are acquiring raw ADC data."""
+    adc_stream: Optional[_ResultSource] = None
+    """Stream to collect raw ADC data."""
+
+    threshold: Optional[float] = None
+    """Threshold to be used for classification of single shots."""
+    angle: Optional[float] = None
+    """Angle in the IQ plane to be used for classification of single shots."""
+    shot: Optional[_Variable] = None
+    shots: Optional[_ResultSource] = None
+    """Variable and stream to collect the classified shots.
+    Used only if a threshold and angle is given.
+    """
+
+    def __post_init__(self):
+        """Create QUA variables needed for single shot classification."""
+        if self.raw_adc:
+            self.adc_stream = declare_stream(adc_trace=True)
+
+        if self.threshold is not None:
+            self.shot = declare(bool)
+            self.shots = declare_stream()
+            self.cos = np.cos(self.angle)
+            self.sin = np.sin(self.angle)
+
+    def save(self):
+        """QUA instruction to save acquired results from variables to streams."""
+        save(self.I, self.I_stream)
+        save(self.Q, self.Q_stream)
+        if self.threshold is not None:
+            save(self.shot, self.shots)
+
+    def classify_shots(self):
```

Comment:
All these functions (this and former) that take nothing and return nothing look a bit weird, especially when you call them (it is not clear to the caller what they are operating on).

This would look nice as a `self.shot` setter (with `shot` being a property), but you need the rhs to be available to the caller.

Reply:
You could think about splitting the two parts:
- the rhs in one function
- the assignment in the setter

Of course:
```py
self.shot = self.classify()
```
still looks a bit weird.

I believe that at the moment, this class is a large context for a computation, more than an *object*.

Reply:
> I believe that at the moment, this class is a large context for a computation, more than an _object_.

Indeed, all these variables are used by the QUA program and the weird functions are part of this program, in fact they only work under a
```py
with program() as prog:
    ...
```
scope, even though this is not entirely clear with the way I organized the code.

I thought that putting them in a class heps with organization and documentation but if you have a better idea let me know.

Reply:
Indeed, I suspected it worked that way.

I believe the class idea should be fine, if there are repeated patterns and assumptions we can encode, but it should be a bit more "transparent": you could open a `program()` context by creating a class instance, and close explicitly (and optionally provide another context manager, to do everything automatically), and the operations should be a bit more composable, taking inputs and returning outputs, keeping a minimal state.
In a sense, I'd turn more in a "wrapper" context, with some pre-encoded behavior.

However, designing this would take a while, and I've limited availability right now, so maybe we could just postpone a bit :)
At the end of the day, I believe the class might be a useful tool to split a part of the workload, we can improve it later on.